### DISCUSSION TITLE
tip: way to use Helpful-Extensions ShapeTracing in  netcore 3.1

### DISCUSSION THREAD

**Discussion by scil** [Lombiq/Helpful-Extensions#80] [Show and tell]
```
mkdir test && cd test
clone https://github.com/Lombiq/Helpful-Extensions.git
# a version in 2021.9
git checkout 4d50142d

# add netcoreapp3.1 and 9.0
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <LangVersion>9.0</LangVersion>
  </PropertyGroup>


cd ..
dotnet add Your.Application reference test\Helpful-Extensions
dotnet sln add test\Helpful-Extensions\Lombiq.HelpfulExtensions.csproj

cd test && clone https://github.com/Lombiq/Helpful-Libraries.git
# a version in 2021.9
git checkout 9c18b777
cd ..
## maybe add netcoreapp3.1 and 9.0
dotnet add Your.Application reference test\Helpful-Libraries\Lombiq.HelpfulLibraries
dotnet sln add test\Helpful-Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj


// comment out the line if error reported
            if (!_hca.HttpContext.IsDevelopment()) return Task.CompletedTask;

```

at last, visit /admin and enable feature of id  `Lombiq.HelpfulExtensions.ShapeTracing` (name: Shape Tracing Helpful Extensions - Lombiq Helpful Extensions)

# 中文/Chinese

// 目前工具只支持net6.0 :  */
// 1. dotnet add package Lombiq.HelpfulLibraries 然后象下面那样添加工程引用和解决方案引用
// 2. 象下面那样手动下载和添加 Helpful-Extensions
// 3. 运行无误，在后台启用特性 Lombiq.HelpfulExtensions.ShapeTracing

// 支持3.1的方法 （2021.10 支持net 3.1 成功）
cd test
clone https://github.com/Lombiq/Helpful-Extensions.git
// 取出 2021.9 的版本
git checkout 4d50142d
// 修改target版本号为3.1；语言为9.0
cd ..
dotnet add Tk.Application.Admin reference test\Helpful-Extensions
dotnet sln add test\Helpful-Extensions\Lombiq.HelpfulExtensions.csproj

cd test && clone https://github.com/Lombiq/Helpful-Libraries.git
// 把 2021.9 的提交取出来
git checkout 9c18b777
cd ..
dotnet add Tk.Application.Admin reference test\Helpful-Libraries\Lombiq.HelpfulLibraries
dotnet sln add test\Helpful-Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj

// 两个项目上设置版本 否则提示某些语法“在 c# 8.0 下不可用”
  <PropertyGroup>
    <LangVersion>9.0</LangVersion>
  </PropertyGroup>

// 这几行报错的话 注释掉 反正不影响使用
            if (!_hca.HttpContext.IsDevelopment()) return Task.CompletedTask;

运行无误，在后台启用特性 Lombiq.HelpfulExtensions.ShapeTracing (name: Shape Tracing Helpful Extensions - Lombiq Helpful Extensions)

**Comment by Piedone** [comment]
Thank you for sharing this! I'm sure this will come handy for people.
